### PR TITLE
Provide variant details on `ResolutionResult`

### PR DESCRIPTION
**Pull request by melix**
### Context

This pull request adds information about the resolved variant on `ResolutionResult`, more precisely through `ResolvedComponentResult`:

`ResolvedVariantResult getVariant();`

A resolved variant result currently consists of:

- a variant display name
- variant attributes

There are 2 possible technical outcomes for what a "variant" is, but they are really the same thing:

- for "non variant aware" components, they consist of the target configuration. So if `project A` resolves to `project B, configuration 'runtime'`, the "variant" is `runtime`. If `project A` resolves to an external Maven component `org:foo:1.0`, scope `compile`, then the variant is `compile`
- for "variant aware" components, the result consists of the selected variant. This will turn particularly useful for Android projects.

It's worth noting that this pull request does **not** improve the dependency insight report or the dependencies report. This should be implemented separately.

### CODE REVIEW

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/result/ResolvedNamedVariantResult.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import org.gradle.api.Named;
+
+/**
+ * Provides information about a named resolved variant.
+ *
+ * @since 4.5
+ *
+ */
+@Incubating
+public interface ResolvedNamedVariantResult extends Named, ResolvedVariantResult {
```
Comment (bigdaz) [inline subprojects/core-api/src/main/java/org/gradle/api/artifacts/result/ResolvedNamedVariantResult.java]: :x: I don't think we need this. Instead, we should add `getDisplayName()` to `ResolvedVariantResult`. We don't want the variant _name_ to become a strong part of our model.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/ComponentResult.java [hunk 2]
```diff
@@ -38,4 +39,8 @@
      * Returns a unique id for this component.
      */
     ComponentIdentifier getComponentId();
+
+    String getVariantName();
```
Comment (bigdaz) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/ComponentResult.java]: :x: Let's add javadoc to our internal APIs

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/result/ResolvedNamedVariantResult.java (hunk 1 above)
Comment (melix) [reply subprojects/core-api/src/main/java/org/gradle/api/artifacts/result/ResolvedNamedVariantResult.java]: Done

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/ComponentResult.java (hunk 2 above)
Comment (melix) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/ComponentResult.java]: Added

**Comment by melix** [comment]
@bigdaz @jjohannes any update on this? Can it be merged?

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/result/DefaultResolvedArtifactResult.java [hunk 3]
```diff
@@ -24,14 +24,20 @@
...
 public class DefaultResolvedArtifactResult implements ResolvedArtifactResult {
+    private static final String UNDEFINED_VARIANT_NAME = "undefined";
+
     private final ComponentArtifactIdentifier identifier;
     private final ResolvedVariantResult variant;
     private final Class<? extends Artifact> type;
     private final File file;
 
     public DefaultResolvedArtifactResult(ComponentArtifactIdentifier identifier, AttributeContainer variantAttributes, Class<? extends Artifact> type, File file) {
+        this(identifier, UNDEFINED_VARIANT_NAME, variantAttributes, type, file);
```
Comment (bigdaz) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/result/DefaultResolvedArtifactResult.java]: This isn't right. If the caller provides the variant attributes, they should also provide the variant display name.
It looks like `ArtifactVisitor.visitArtifact()` and `visitFile` could either take a `ResolvedVariant` in place of `AttributeContainer` or have an additional 'variantName' parameter.

For each caller of `ArtifactVisitor.visitArtifact()` you should be able to provide a reasonable display name. Certainly `ArtifactBackedResolvedVariant` already has a 'displayName' property that might suffice.

**Comment by melix** [comment]
Alright I have made the variant name available in `visitArtifact` and friends. There's however one place where I didn't know what to use: https://github.com/melix/gradle/blob/cc3233d77fd36bc885e189b262e3c8a161d46e22/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/artifact/LocalFileDependencyBackedArtifactSet.java#L85

This is for "local file dependencies". I'm just using "local file" as a variant name in that case.

**Pull request merged**